### PR TITLE
SLING-10952 - Adding support for whitespace characters in group names

### PR DESCRIPTION
**Pull request by klcodanr**
Goal here is to be able to define repoinit statements for groups loaded from external sources that contain whitespace in the name. Allows group names to either be quoted or standard strings. Note this only adds creation / deletion support for groups, but theoretically you could update properties and update membership of other types of authorizables with whitespace in the name.

### CODE REVIEW

File: src/main/java/org/apache/sling/repoinit/parser/impl/AuthorizableIdUtil.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.sling.repoinit.parser.impl;
+
+import java.util.List;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+public class AuthorizableIdUtil {
```
Comment (anchela) [inline src/main/java/org/apache/sling/repoinit/parser/impl/AuthorizableIdUtil.java:23]: i find the name of this class a bit unfortunate as it is used both for parsing authorizable ids as well as for principal names.

second, i would suggest to write a dedicate unit test for the methods of this utility class

File: src/main/java/org/apache/sling/repoinit/parser/impl/AuthorizableIdUtil.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    /**
+     * Gets the authorizable ID handling cases where the value should be quoted
+     * (e.g. the authorizable id is not just an alphanumeric string)
+     * 
+     * @param authorizableId the authorizable ID to get
+     * @return the (potentially quoted) authorizable id
+     */
+    public static final String forRepoInitString(String authorizableId) {
```
Comment (anchela) [inline src/main/java/org/apache/sling/repoinit/parser/impl/AuthorizableIdUtil.java:38]: same here, sometimes a principal name is passed which isn't the same as the authorizable id.
i would also suggest to use notnull/nullable annotations for the param and the return value and adjust the description accordingly

File: src/main/java/org/apache/sling/repoinit/parser/impl/AuthorizableIdUtil.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    /**
+     * Gets the authorizable IDs handling cases where the value should be quoted
+     * (e.g. the authorizable id is not just an alphanumeric string)
+     * 
+     * @param authorizableIds the authorizable IDs to get
+     * @return the list of (potentially quoted) authorizable ids
+     */
+    public static final List<String> forRepoInitString(List<String> authorizableIds) {
```
Comment (anchela) [inline src/main/java/org/apache/sling/repoinit/parser/impl/AuthorizableIdUtil.java:49]: same as above

File: src/main/java/org/apache/sling/repoinit/parser/operations/AclGroupBase.java [hunk 4]
```diff
@@ -40,7 +41,7 @@
     private final List<String> aclOptions;
     
     protected AclGroupBase(List<AclLine> lines) {
-        this(lines,new ArrayList<String>());
+        this(lines,new ArrayList<>());
```
Comment (anchela) [inline src/main/java/org/apache/sling/repoinit/parser/operations/AclGroupBase.java:44]: IMHO this is an unrelated change. if that is true please move it to a separate pull request

File: src/main/java/org/apache/sling/repoinit/parser/operations/SetAclPaths.java [hunk 5]
```diff
@@ -41,6 +41,7 @@ public SetAclPaths(List<String> paths,List<AclLine> lines, List<String> aclOptio
         this.paths = Collections.unmodifiableList(paths);
     }
     
+    @Override
```
Comment (anchela) [inline src/main/java/org/apache/sling/repoinit/parser/operations/SetAclPaths.java:44]: again, IMHO this change is unrelated to the task at hand. i would prefer if it would got in a separate 'minor cleanup' PR with a dedicated ticket.

File: src/main/java/org/apache/sling/repoinit/parser/operations/SetAclPrincipalBased.java [hunk 6]
```diff
@@ -33,14 +34,15 @@
     private final List<String> principals;
 
     public SetAclPrincipalBased(List<String> principals, List<AclLine> lines) {
-        this(principals,lines,new ArrayList<String>());
+        this(principals,lines,new ArrayList<>());
```
Comment (anchela) [inline src/main/java/org/apache/sling/repoinit/parser/operations/SetAclPrincipalBased.java:37]: see above

File: src/main/java/org/apache/sling/repoinit/parser/operations/SetAclPrincipalBased.java [hunk 7]
```diff
@@ -33,14 +34,15 @@
...
     }
 
     public SetAclPrincipalBased(List<String> principals, List<AclLine> lines, List<String> aclOptions) {
         super(lines,aclOptions);
         this.principals = Collections.unmodifiableList(principals);
     }
 
+    @Override
```
Comment (anchela) [inline src/main/java/org/apache/sling/repoinit/parser/operations/SetAclPrincipalBased.java:45]: see above

File: src/main/java/org/apache/sling/repoinit/parser/operations/SetAclPrincipals.java [hunk 8]
```diff
@@ -33,14 +34,15 @@
     private final List<String> principals;
     
     public SetAclPrincipals(List<String> principals, List<AclLine> lines) {
-        this(principals,lines,new ArrayList<String>());
+        this(principals,lines,new ArrayList<>());
```
Comment (anchela) [inline src/main/java/org/apache/sling/repoinit/parser/operations/SetAclPrincipals.java:37]: see above

File: src/main/java/org/apache/sling/repoinit/parser/operations/SetAclPrincipals.java [hunk 9]
```diff
@@ -33,14 +34,15 @@
...
     }
 
     public SetAclPrincipals(List<String> principals,List<AclLine> lines, List<String> aclOptions) {
         super(lines,aclOptions);
         this.principals = Collections.unmodifiableList(principals);
     }
 
+    @Override
```
Comment (anchela) [inline src/main/java/org/apache/sling/repoinit/parser/operations/SetAclPrincipals.java:45]: see above

File: src/main/javacc/RepoInitGrammar.jjt [hunk 10]
```diff
@@ -771,13 +771,21 @@ void removeFromGroupStatement(List<Operation> result) :
...
     <FROM> <GROUP>
-    group = <STRING>
+    group = authorizableId()
 
     {
         result.add(new RemoveGroupMembers(members, group.image));
     }
 }
 
+Token authorizableId() :
```
Comment (anchela) [inline src/main/javacc/RepoInitGrammar.jjt:781]: same as with the utility class. the name of the method is IMO not correct as it is used both for principal names and for authorizable ids.

**Review by anchela — CHANGES_REQUESTED** [review]
@klcodanr , i am fine with overall direction but have requested a few minor changes.

**Comment by klcodanr** [comment]
@anchela Good points.

Re:Naming, would you have any suggestion on the name for this class / the related tokens and methods given that it is used for Authorizable ID's and Principal Names?

It feels like it may be easiest / most logical to call it PrincipalName (e.g. PrincipalNameUtil / principalName / principalName()) as long as that's not confusing since it an also be an Authorizable ID. 

Would you agree?

Re:the other changes I'll make those and send an update. 

Thanks!

**Comment by klcodanr** [comment]
I was thinking about this more and the name is misleading about what the purpose of the token is. While it is used for principal names and authorizable ids, it's really a "quotable string" i.e. a string that can either be quoted which allows whitespace or not which does not allow whitespace.

**Comment by anchela** [comment]
@klcodanr , i agree with that analysis. changing the method name to what it actually is instead of what it is used for makes sense to me. thanks a lot.

**Comment by klcodanr** [comment]
@anchela I believe I made all the changes requested. However I do see [SLING-10996](https://issues.apache.org/jira/browse/SLING-10996). Should instead this be aligned with the allowed characters in a JCR local name?

I'd like to get this in, but I'd prefer to do it correctly vs having a _mostly_ though not entirely correct implementation.

**Comment by anchela** [comment]
@klcodanr , i fear SLING-10996 will be a bigger thing to do. i would not include it in this improvement here, which is primarily about allow for white space in strings like user/group names.
